scripts/test_lib.sh

```diff
@@ -166,7 +166,7 @@ function run_for_module {
 }
 
 function module_dirs() {
-  echo "api pkg client/pkg client/internal/v2 client/v3 server etcdutl etcdctl tests ."
+  echo "api pkg client/pkg client/internal/v2 client/v3 server etcdutl etcdctl tests tools/rw-heatmaps tools/testgrid-analysis ."
```

Comment:
There is a function below called `modules` on line `181`.   Do we also need to update that @ahrtr?

Reply:
I think yes.

Reply:
@ahrtr, @jmhbnz, adding these to `modules`, breaks generating the BOM. This diff partially fixes the issue:

```diff
diff --git a/dummy.go b/dummy.go
index 48db9af93..ee4ff91e6 100644
--- a/dummy.go
+++ b/dummy.go
@@ -18,8 +18,10 @@ package main_test
 // module (e.g. for sake of 'bom' generation).
 // Thanks to this 'go mod tidy' is not removing that dependencies from go.mod.
 import (
-       _ "go.etcd.io/etcd/client/v2"                // keep
-       _ "go.etcd.io/etcd/etcdctl/v3/ctlv3/command" // keep
-       _ "go.etcd.io/etcd/etcdutl/v3/etcdutl"       // keep
-       _ "go.etcd.io/etcd/tests/v3/integration"     // keep
+       _ "go.etcd.io/etcd/client/v2"                      // keep
+       _ "go.etcd.io/etcd/etcdctl/v3/ctlv3/command"       // keep
+       _ "go.etcd.io/etcd/etcdutl/v3/etcdutl"             // keep
+       _ "go.etcd.io/etcd/tests/v3/integration"           // keep
+       _ "go.etcd.io/etcd/tools/rw-heatmaps/v3/cmd"       // keep
+       _ "go.etcd.io/etcd/tools/testgrid-analysis/v3/cmd" // keep
 )
diff --git a/go.mod b/go.mod
index 9d9aae976..ca4d6ebd4 100644
--- a/go.mod
+++ b/go.mod
@@ -14,6 +14,8 @@ replace (
        go.etcd.io/etcd/pkg/v3 => ./pkg
        go.etcd.io/etcd/server/v3 => ./server
        go.etcd.io/etcd/tests/v3 => ./tests
+       go.etcd.io/etcd/tools/rw-heatmaps/v3 => ./tools/rw-heatmaps
+       go.etcd.io/etcd/tools/testgrid-analysis/v3 => ./tools/testgrid-analysis
 )
 
 require (
@@ -33,6 +35,8 @@ require (
        go.etcd.io/etcd/pkg/v3 v3.6.0-alpha.0
        go.etcd.io/etcd/server/v3 v3.6.0-alpha.0
        go.etcd.io/etcd/tests/v3 v3.6.0-alpha.0
+       go.etcd.io/etcd/tools/rw-heatmaps/v3 v3.6.0-alpha.0
+       go.etcd.io/etcd/tools/testgrid-analysis/v3 v3.6.0-alpha.0
        go.etcd.io/raft/v3 v3.6.0-alpha.0
        go.uber.org/zap v1.27.0
        golang.org/x/time v0.5.0
@@ -41,21 +45,31 @@ require (
 )
```

Adding them to `dummy.go` so that they are added to `go.mod`. However, after doing that, the error from building the BOM looks random (fails at `pkg/expect/expect.go` and `tests/robustness/report/client.go`). I'm not sure if you have any hints of what's wrong:

```
2024/04/12 20:24:53 could not list go.etcd.io/etcd/api/v3/... go.etcd.io/etcd/pkg/v3/... go.etcd.io/etcd/client/pkg/v3/... go.etcd.io/etcd/client/v2/... go.etcd.io/etcd/client/v3/... go.etcd.io/etcd/server/v3/... go.etcd.io/etcd/etcdutl/v3/... go.etcd.io/etcd/etcdctl/v3/... go.etcd.io/etcd/tests/v3/... go.etcd.io/etcd/tools/rw-heatmaps/v3/
... go.etcd.io/etcd/tools/testgrid-analysis/v3/... go.etcd.io/etcd/v3/... dependencies: 'go list go.etcd.io/etcd/api/v3/... go.etcd.io/etcd/pkg/v3/... go.etcd.io/etcd/client/pkg/v3/... go.etcd.io/etcd/client/v2/... go.etcd.io/etcd/client/v3/... go.etcd.io/etcd/server/v3/... go.etcd.io/etcd/etcdutl/v3/... go.etcd.io/etcd/etcdctl/v3/... go.e
tcd.io/etcd/tests/v3/... go.etcd.io/etcd/tools/rw-heatmaps/v3/... go.etcd.io/etcd/tools/testgrid-analysis/v3/... go.etcd.io/etcd/v3/...' failed with:
pkg/expect/expect.go:32:2: missing go.sum entry for module providing package github.com/creack/pty (imported by go.etcd.io/etcd/pkg/v3/expect); to add:
        go get go.etcd.io/etcd/pkg/v3/expect@v3.6.0-alpha.0
tests/robustness/report/client.go:27:2: missing go.sum entry for module providing package github.com/anishathalye/porcupine (imported by go.etcd.io/etcd/tests/v3/robustness/model); to add:
        go get go.etcd.io/etcd/tests/v3/robustness/model@v3.6.0-alpha.0
```

Reply:
Have any updates for this?

Reply:
Sorry, did not get time to dig into this. But one thing for sure is that we don't want the `tools/*` to break any production packages.

For simplicity & safety, propose to take care of (e.g. `go mod tidy; gofmt, etc`) all the packages under the tools/* separately, for example, we can extend the existing `scripts/fix.sh`.

Reply:
@ivanvc  could you guide the contributor to finish this? cc @ArkaSaha30

Reply:
Hi @ahrtr, I meant to get back to this but was busy with other tasks. I'll take a stab at this.

Reply:
By the way, @ahrtr, do we want the BOM to be updated with the dependencies from the tools sub-projects? I think that's what's problematic right now.

Reply:
> By the way, @ahrtr, do we want the BOM to be updated with the dependencies from the tools sub-projects?

Currently the BOM doesn't include any dependencies in tools. So I think we just need to keep it as it's.

Reply:
@ahrtr, I've carefully reviewed the bash scripts inside `scripts/`. Although we could create some new functions to declare the tool directories/modules and break down when we should run something on them and when not, I don't think it makes sense.

If we keep the tool modules out of the `modules` function/variable, then it means that the only tasks that won't run on them are:
1. BOM
2. Tests coverage

The rest of the tasks use `module_dirs`, which the contributor initially did. My suggestion would be to add a comment to the `modules` function specifying that the tools shouldn't be in there.

Do we care about test coverage (2.) for the tools?

cc. @ArkaSaha30

Reply:
Thanks @ivanvc for the summary. 

- For BOM (bill-of-materials), as mentioned above, current the [BOM](https://github.com/etcd-io/etcd/blob/main/bill-of-materials.json) doesn't include any dependencies used by [tools](https://github.com/etcd-io/etcd/tree/main/tools). I was thinking to keep it as it's, because the dependences in tools are only supposed to be used in **dev** environment. In other words, any dependencies used by tools do not affect the security or license concerns of the etcd running in production environment. I suggest to keep it as it's unless someone (e.g. from the [K8s SRC](https://github.com/kubernetes/committee-security-response)) clearly requests/challenges it. (The other option is to create a separate list of BOM for the tools.) Could you raise a discussion in the sig-etcd channel to get feedback and CC the K8s SRC members?
- For the test coverage, it isn't a big deal.


Note the purpose of this PR is to ensure the scripts/fix.sh can also automatically run `go mod tidy` and `gofmt` for all the packages under tools, so that we don't need to manually take care of them separately.

Reply:
I'll start a thread in #sig-etcd. `scripts/fix.sh` will run on the tools subdirectories, as it uses `module_dirs`.

@tico88612, sorry for the back-and-forth. But we're now aligned on how to solve the issue. 
1. Can you please revert commit 8d38d1f855fd8f892cbe7896fb0823cd6fb21413
2. Can you apply the following patch after reverting the previously mentioned commit?
   ```diff
   diff --git a/scripts/test_lib.sh b/scripts/test_lib.sh
   index 550c718d3..be5194fb1 100644
   --- a/scripts/test_lib.sh
   +++ b/scripts/test_lib.sh
   @@ -178,6 +178,9 @@ function maybe_run() {
      fi
    }
 
   +# modules
   +# returns the list of all modules in the project, not including the tools,
   +# as they are not considered to be added to the bill for materials.
    function modules() {
      modules=(
        "${ROOT_MODULE}/api/v3"
   ```

Please let me know if you have any follow-up questions. If you don't have the capacity to work on this anymore, I can start a new PR and co-author the commits you made. Thank you so much for your contribution.